_posts/2021-05-18-managing-react-application-state-with-context.md

```diff
@@ -0,0 +1,177 @@
+
+```javascript
+// countProvider.tsx
+function CountReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case "increment":
+      return {
+        ...state,
+        count: state.count + 1,
+      };
+    case "decrement":
+      return {
+        ...state,
+        count: state.count - 1,
+      };
+    case "setValue": {
+      return {
+        ...state,
+        count: action.payload.value,
+      };
+    }
+    default:
+      throw Error(`Unhandled action type: ${action!.type}`);
+  }
+}
+```
+
+Next, I set up Context for both the `State` and `Dispatch` and provide a function component that determines how the `Reducer` we created above will be used. I also provide a way of setting the initial state which aids in testability since you can establish a given state in your tests before performing actions to see how the system behaves.
+
+```javascript
+// countProvider.tsx
+const CountStateContext = createContext<State | undefined>(undefined);
+const CountDispatchContext = createContext<Dispatch | undefined>(undefined);
+
+const CountProvider: FC<{ init?: Partial<State> }> = ({ children, init }) => {
+  const initialState: State = {
+    count: 0,
+    ...init,
+  };
+
+  const [state, dispatch] = useReducer(CountReducer, initialState);
+
+  return (
+    <CountStateContext.Provider value={state}>
+      <CountDispatchContext.Provider value={dispatch}>
+        {children}
+      </CountDispatchContext.Provider>
+    </CountStateContext.Provider>
+  );
+};
+```
+
+I then wrap our App in the `CountProvider` component, ensuring that all of the components that rely on the `State` or will perform `Action`s on that state are encapsulated as children of this component.
+
+```javascript
+// App.tsx
+function App() {
+  return (
+    <CountProvider>
+      <CountDisplay />
+      <CountButtons />
+      <CountInput />
+    </CountProvider>
+  );
+}
+```
+
+Finally, I provide some convenience functions which allow us to easily access the our `State` and `Dispatch` actions within components along with providing some error feedback if we are trying to access them outside of our `CountProvider`.
+
+```javascript
+// countProvider.tsx
+export function useCount(): [State, Dispatch] {
+  const state = useContext(CountStateContext);
+  const dispatch = useContext(CountDispatchContext);
+  if (state === undefined || dispatch === undefined) {
+    throw new Error("useCount must be used within a CountProvider");
+  }
+  return [state, dispatch];
```

Comment:
Ahh this is new 👍 , previously we would create 2 helper functions, each for State and Dispatch. Hmmm.. after some thought I'm wondering about scenarios where you've got a component which only requires the Dispatch to make a change, but doesn't care about State updates (i.e. you've got a button to make a change), doesn't this mean it'll now get disrupted upon State changes? I wonder if this is why the original implementation created 2 functions?

Reply:
I think this is just personal preference so instead of having both `useCountState` and `useCountDispatch` its just more less noisy to just have `useCount` and in the case where we are just dispatching actions we can use
```javascript
const [, dispatch] = useCount();
```
or for just state
```javascript
const [{ count }] = useCount();
```

Reply:
In either case, whenever the state changes, the child components will update unless you `memo` them